### PR TITLE
Improve hypervisor ranking

### PR DESCRIPTION
**Pull request by hasegeli**

### CODE REVIEW

**Comment by HerrSpace** [comment]
Can one of the admins verify this patch?

File: igvm/commands.py [hunk 1]
```diff
@@ -105,26 +105,26 @@ def disk_set(vm_hostname, size, ignore_reserved=False):
...
-    vm.server_obj['disk_size_gib'] = new_size_gib
-    vm.server_obj.commit()
+    vm.dataset_obj['disk_size_gib'] = new_size_gib
+    vm.dataset_obj.commit()
 
 
 @with_fabric_settings
 def vm_build(vm_hostname, localimage=None, nopuppet=False, postboot=None,
-             ignore_reserved=False, balance_config=None):
+             ignore_reserved=False):
```
[the remark this answers is not in the document]
Comment (HerrSpace) [reply igvm/commands.py]: Also we need a new concept, how to clear out a group of HVs, w/o igvm migrating VMs back to them. Before we could just add a balance rule not to choose a HV with some serveradmin attribute value, e.g. datacenter even if that HV has the appropriate vlan_network.

File: igvm/balance/constraints.py [hunk 2]
```diff
@@ -44,25 +44,19 @@ def __call__(self, vm, hv):
         return hv.get_memory_free(fast=True) > vm.get_memory()
 
 
-class EnsureFunctionDistribution(object):
-    """Game World / Function Distribution
+class NoVMWithSameAttributes(object):
+    """Check the hypervisor has another VM with the same attributes"""
```
[the remark this answers is not in the document]
Comment (hasegeli) [reply igvm/balance/constraints.py]: Changing...

File: igvm/balance/constraints.py [hunk 3]
```diff
@@ -44,25 +44,19 @@ def __call__(self, vm, hv):
...
-        """
-        return all(
-            hv_vm.hostname == vm.hostname or
-            hv_vm.get_identifier() != vm.get_identifier()
-            for hv_vm in hv.get_vms()
-        )
+        for other_vm in hv.get_vms():
+            if other_vm.hostname == vm.hostname:
+                continue
+            if all(other_vm[a] == vm[a] for a in self.attributes):
```
[the remark this answers is not in the document]
Comment (hasegeli) [reply igvm/balance/constraints.py]: They are not though.  Serveradmin guarantees same set of attributes to exist on the same `servertype`.  They will be `None` when not set.

File: igvm/balance/rules.py [hunk 4]
```diff
@@ -58,3 +58,23 @@ def __call__(self, vm, hv):
...
+
+        # Nothing to do if there are no other VMs
+        if not other_vms:
+            return 0.0
+
+        weight = 1.0 / len(other_vms)
+
+        return sum(
+            weight for v in other_vms if v['xen_host'] == hv['hostname']
+        )
```
[the remark this answers is not in the document]
Comment (hasegeli) [reply igvm/balance/rules.py]: Yes

File: igvm/balance/models.py [hunk 5]
```diff
@@ -4,18 +4,17 @@
 """
 
 import adminapi.api
+from adminapi.dataset import query
```
[the remark this answers is not in the document]
Comment (hasegeli) [reply igvm/balance/models.py]: I do on the newer version.

File: igvm/balance/constraints.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (HerrSpace) [reply igvm/balance/constraints.py]: Sure, I was talking about a hypothetical other serveradmin instance, operated by people who might might not have a use for these attributes on their VMs.

File: igvm/balance/constraints.py [hunk 6]
```diff
@@ -86,11 +83,11 @@ def __call__(self, vm, hv):
 
         # If we don't try to migrate a master db we can safely ignore further
         # evaluation.
-        if vm['game_world'] != 0 or vm['function'] not in self.db_types:
+        if vm['game_world'] != 0 or 'db' not in vm['function']:
             return True
 
         return all(
-            hv_vm['game_world'] != 0 or hv_vm['function'] not in self.db_types
+            hv_vm['game_world'] != 0 or 'db' not in hv_vm['function']
```
[the remark this answers is not in the document]
Comment (hasegeli) [reply igvm/balance/constraints.py]: I re-implemented this in more generic way.

File: igvm/commands.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hasegeli) [reply igvm/commands.py]: I made the constraints not hard-filtering.

**Comment by HerrSpace** [comment]
If you switch the HV1 from 'aw-hv-053.ndco.ig.local' to 'aw-hv-058.ndco.ig.local' the tests would have a chance to pass btw, or we can just merge Kajetans MR first.

File: igvm/vm.py [hunk 7]
```diff
@@ -545,6 +549,31 @@ def get_best_hypervisor(self, hv_states=['online']):
...
+=======
+            # Free disk size and memory are done in fast mode which might
+            # return wrong or orphaned values so to be sure we really have
+            # enough we check them here again.
+            disk_free = hypervisor.get_disk_free() / 1024.0
+            if float(disk_free) > float(self.dataset_obj['disk_size_gib']):
+                memory_free = hypervisor.get_memory_free()
+                if memory_free > float(self.dataset_obj['memory']):
+                    return hypervisor['hostname']
+>>>>>>> 402bef9... balance: Remove ServeradminCache
```
Comment (HerrSpace) [inline igvm/vm.py]: Unresolved conflict.

File: igvm/vm.py [hunk 8]
```diff
@@ -534,46 +534,15 @@ def get_best_hypervisor(self, hv_states=['online']):
...
-                if memory_free > float(self.dataset_obj['memory']):
-                    return hypervisor['hostname']
->>>>>>> 402bef9... balance: Remove ServeradminCache
+            # Free disk size and memory are done in fast mode on ranking.
+            # They might be wrong, so to be sure we really have enough,
+            # we check them here again.
+            if hypervisor.get_disk_free() <= self.dataset_obj['disk_size_gib']:
+                continue
+            if hypervisor.get_memory_free() <= self.dataset_obj['memory']:
+                continue
```
Comment (HerrSpace) [inline igvm/vm.py]: get_disk_free() and get_memory_free() can now return None if the libvirt connection failed, which will raise a TypeError when compared with ints via the `<=` operator in python3.

File: igvm/host.py [hunk 9]
```diff
@@ -32,35 +36,13 @@ def get_server(hostname, servertype):
...
-        )
-
-    return server
+    return Query(
+        {
+            'servertype': servertype,
+            'hostname': Or(*conditions),
+        },
+        HYPERVISOR_ATTRIBUTES if servertype == 'hypervisor' else VM_ATTRIBUTES
+    ).get()
```
Comment (HerrSpace) [inline igvm/host.py]: Why even have this function at all?

File: igvm/hypervisor_preferences.py [hunk 10]
```diff
@@ -67,3 +71,48 @@ def __init__(self, threshold):
...
+    """
+
+    def __call__(self, vm, hv):
+        return 100.0 - (hv.dataset_obj['cpu_util_pct'] or 100.0)
+
+
+class CPUOverAllocation(object):
+    """CPU Over Allocation
+
+    Rate the CPU over allocation of the target hypervisor and if it better or
```
Comment (HerrSpace) [inline igvm/hypervisor_preferences.py]: s/if it better/if it is better/

File: igvm/hypervisor_ranking.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        'vlan_networks': vm.dataset_obj['route_network'],
+        'state': Any(*hypervisor_states),
+    }
+
+    log.info('Evaluating hypervisors...')
+
+    return (r.hypervisor for r in sorted((
+        HypervisorRanking(vm, Hypervisor(o))
+        for o in Query(filters, HYPERVISOR_ATTRIBUTES)
+    ), key=cmp_to_key(HypervisorRanking.cmp), reverse=True))
```
Comment (HerrSpace) [inline igvm/hypervisor_ranking.py]: `key=cmp_to_key(HypervisorRanking.cmp),` is not the easiest construct to get.

File: igvm/host.py (hunk 9 above)
Comment (hasegeli) [reply igvm/host.py]: I don't like this either.  The caller should always make the query before initializing the objects.  We can refactor this later.

File: igvm/hypervisor_ranking.py (hunk 11 above)
Comment (hasegeli) [reply igvm/hypervisor_ranking.py]: I will comment it.

File: igvm/hypervisor_preferences.py (hunk 10 above)
Comment (hasegeli) [reply igvm/hypervisor_preferences.py]: Fixed.

File: igvm/vm.py (hunk 7 above)
Comment (hasegeli) [reply igvm/vm.py]: Fixed.

File: igvm/vm.py (hunk 8 above)
Comment (hasegeli) [reply igvm/vm.py]: The methods of the `Hypervisor` class don't really return `None`.

Comment (HerrSpace) [reply igvm/vm.py]: Well the try, finally block suggests that it can if ighv.free_vm_memory() raises an Exception.

**Pull request merged**